### PR TITLE
ready for review

### PR DESCRIPTION
**Pull request by skeydan**
Hi @kevinykuo, @batpigandme, could you please take a look? Thanks!

### CODE REVIEW

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+    self_contained: false
+preview: images/thumb.png
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+So before we jump into the technicalities, -- yes, that's what this post is too: _Un hommage à_ McElreath who wrote one of most intriguing books on Bayesian (or should we just say - scientific?) modeling we're aware of. If you haven't read _Statistical Rethinking_, and are interested in modeling, you might definitely want to check it out. In this post, we're not going to try to re-tell the story: Our clear focus will instead be a demonstration of how to do MCMC with [tfprobability](https://rstudio.github.io/tfprobability/).
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: I think the " -- yes, that's what this post is too" is a little hard to understand with the sentence construct. I think it's implied by the first part of the sentence, so you can probably cut it.

Comment (batpigandme) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Offset instead with commas

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+
+Concretely, this post has two parts. The first is a quick overview of how to use `tfd_joint_sequential_distribution` to construct a model, and then sample from it using Hamiltonian Monte Carlo. This part could be consulted for quick code look-up, or as a frugal template of the whole process.
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Link to `tfd_join_sequential_distribution` function reference, if you have it?

Comment (batpigandme) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: "This part could be…" I think you can use _can_ here (can be consulted…)

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+```
+'data.frame':	48 obs. of  5 variables:
+ $ density : int  10 10 10 10 10 10 10 10 10 10 ...
+ $ pred    : Factor w/ 2 levels "no","pred": 1 1 1 1 1 1 1 1 2 2 ...
+ $ size    : Factor w/ 2 levels "big","small": 1 1 1 1 2 2 2 2 1 1 ...
+ $ surv    : int  9 10 7 10 9 9 10 9 4 9 ...
+ $ propsurv: num  0.9 1 0.7 1 0.9 0.9 1 0.9 0.4 0.9 ...
+ ```
+ 
+The task is modeling survivor counts among tadpoles, where tadpoles are held in tanks of different sizes (equivalently, different numbers of inhabitants). Each row in the dataset describes one tank, with its initial count of inhabitants (`density`) and number of survivors (`surv`).
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: So are there two variables here: tank size, and number of inhabitants?

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+Note how the first distribution is unconditional, while the second depends on the first. Note too how the second has to be wrapped in `tfd_independent` to avoid wrong broadcasting. (This is an aspect of `tfd_joint_distribution_sequential` usage that deserves to be documented more systematically, which is surely going to happen.^[As of today, lots of useful information is available in [Modeling with JointDistribution](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Modeling_with_JointDistribution.ipynb) and [Multilevel Modeling Primer](https://github.com/tensorflow/probability/blob/master/tensorflow_probability/examples/jupyter_notebooks/Multilevel_Modeling_Primer.ipynb), but some experimentation may needed to adapt the -- numerous! -- examples to your needs.] Just think that this functionality was added to TFP `master` only three weeks ago!)
+
+As an aside, the model specification here ends up shorter than in _Stan_ as `tfd_binomial` optionally takes logits as parameters.
+
+As with every TFP distribution, a quick functionality check is sampling from the model:
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: maybe "…you can do a quick functionality check by sampling from the model:"?

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+Tensor("MultivariateNormalDiag/sample/affine_linear_operator/forward/add:0",
+shape=(2, 48), dtype=float32)
+
+[[2]]
+Tensor("IndependentJointDistributionSequential/sample/Beta/sample/Reshape:0",
+shape=(2, 48), dtype=float32)
+```
+
+You might be wondering why we are not using eager execution, something we normally like to do.
+The reason is that as of today, the sampling code must be executed on the graph in order to be run in parallel -- which definitely is something we'll want to do when with a set of Markov chains. In Python, this can be achieved by annotating the sampling function with `@tf.function`, a construct that as of this writing is not available from R.
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: "as of today" gets offset by commas on both sides

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+
+Now let's see how we can sample from this model using Hamiltonian Monte Carlo.
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Comma after Now

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+) %>%
+  mcmc_simple_step_size_adaptation(
+    target_accept_prob = 0.8,
+    num_adaptation_steps = n_burnin
+  )
+```
+
+We then run the sampler, passing in an initial state. If we want to run $n$ chains, that state has to be of length $n$, for every parameter in the model (here we have just one).
+
+The sampling function, `mcmc_sample_chain`, may optionally be passed a `trace_fn` that tells TFP which kinds of meta information to save. Here we save acceptance ratios and step sizes.
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Again, I try to link to function reference on first mention, but definitely your call!

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+
+plot_posteriors(mcmc_trace[[3]], dim(mcmc_trace[[3]])[3])
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/posteriors.png")
+```
+
+Now let's see the corresponding posterior means and highest posterior density intervals.
+(The below code includes the hyperpriors in `summary` as we'll want to display a complete _precis_-like output soon.)
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Super minor, but I think précis gets an accent aigu over the e!

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+   coord_flip() +  
+   xlab("") + ylab("post. mean and HPDI") +
+   theme_minimal() 
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/forest_plot.png")
+```
+
+Now for an equivalent to _precis_. We already computed means, standard deviations and the HPDI interval.
```
Comment (batpigandme) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: It now occurs to me you could be referring to something technical I don't know about called _precis_, but if just the word, I use an accent. Looks like it's dealer's choice in U.S. English, though.

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Oh yeah it's pretty convoluted... I'm now trying to simply say 

> Before we jump into the technicalities: This post is, of course, dedicated to McElreath who 

That should be at least a bit better right?

Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: sorry I need to ask, what does "Offset instead with commas" refer to exactly?

Comment (batpigandme) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Yes, better! It means use commas around instead so "will, instead, be demonstrated" (sorry if that's not verbatim, I'm on my phone now)

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: right & right!

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 1 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: oh, got it! ... and sorry for having you do this during _staycation_ !!

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: the dataset has several, but only these matter for the current task

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: yep

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 7 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: totally makes sense, esp. with such a new pkg...

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd (hunk 9 above)
Comment (skeydan) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: yeah exactly, it's been "coined" by McElreath and he writes it without the accent ... (((normally I'm snobbish enough to leave them in when I write English && use some French word, but here really I shouldn't ;-)))

**Comment by skeydan** [comment]
Thanks a lot Mara, that's very helpful!! I _do_ notice I've added some  commas && not removed any, which is funny because normally I think Germans try to use fewer commas when writing in English then they're used to ;-)

**Comment by batpigandme** [comment]
Always a pleasure. I'm definitely a comma-advocate.

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 10]
```diff
@@ -0,0 +1,681 @@
...
+And here it is with TFP:
+
+```{r}
+m2 <- tfd_joint_distribution_sequential(
+  list(
+    # a_bar, the prior for the mean of the normal distribution of per-tank logits
+    tfd_normal(loc = 0, scale = 1.5),
+    # sigma, the prior for the variance of the normal distribution of per-tank logits
+    tfd_exponential(rate = 1),
+    # normal distribution of per-tank logits
```
Comment (kevinykuo) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: I'd make a note here that this refers to `l` in the next one in the chain

Comment (kevinykuo) [reply _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: Oh I see you elaborate on this later, but still would be nice to follow what you did above for `a_bar` and `sigma`

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 11]
```diff
@@ -0,0 +1,681 @@
...
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/traceplot_sigma.png")
+```
+
+```{r, eval=TRUE, echo=FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/traceplot_a_1.png")
+```
+
+
+How about the posterior distributions of the parameters, first and foremost, the varying intercepts `a_1` - `a_48`?
```
Comment (kevinykuo) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: ... instead of - ?

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 12]
```diff
@@ -0,0 +1,681 @@
...
+    theme(legend.position = "none",
+          axis.title = element_blank(),
+          axis.text = element_blank(),
+          axis.ticks = element_blank())
+    
+}
+
+plot_posteriors <- function(sample_array, num_params) {
+  plots <- vector(mode = "list", length = num_params)
+  for (i in 1:num_params) {
```
Comment (kevinykuo) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: can we map here instead of looping?

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 13]
```diff
@@ -0,0 +1,681 @@
...
+  arrange(as.integer(str_sub(key, 6))) %>%
+  mutate(key = c("a_bar", "sigma", paste0("a_", 1:48)))
+
+summary <- means %>% 
+  inner_join(sds, by = "key") %>% 
+  inner_join(hpdis_lower, by = "key") %>%
+  inner_join(hpdis_upper, by = "key")
+
+
+summary %>%
```
Comment (kevinykuo) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: y axis labels are overlapping

File: _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd [hunk 14]
```diff
@@ -0,0 +1,681 @@
...
+  tf$reshape(list(2000L, 50L))
+
+# diagnostics, also as above
+is_accepted_ <- res$trace[[1]]
+step_size_ <- res$trace[[2]]
+
+# effective sample size
+# again we use tensorflow to get conveniently shaped outputs
+ess_ <- mcmc_effective_sample_size(mcmc_trace) 
+ess_ <-tf$concat(
```
Comment (kevinykuo) [inline _posts/2019-05-03-tadpoles-on-tensorflow/tadpoles-on-tensorflow.Rmd]: space after `<-`

**Comment by batpigandme** [comment]
Oh, if @kevinykuo doesn't understand the HMC stuff either (new to me, too), then, at the very least I'd link out to what you consider to be a good primer!

**Comment by skeydan** [comment]
Hi Kevin,

thanks a lot, I'm gonna adapt code/plots accordingly!

> Might be worth spending a paragraph or two talking about what the HMC is doing for folks not familiar with it.

I think this is too much a world of its own to introduce in a pragraph, which is why I'm saying I'm not going to try to "re-tell the story". Basically, this is very much a post for "fans", people who've read the book, similar in a way to Joe's (https://rviews.rstudio.com/2018/04/23/on-first-meeting-greta/). 
But I could do what @batpigandme suggests, add a link to some other introduction in a footnote.

In general I think this blog is like RViews in a way that not every post is for everyone, depending on interest.
